File: web-admin/src/services/auth.service.ts
/**
 * Service d'authentification
 * Gère la connexion/déconnexion des utilisateurs
 */

import {
  signInWithEmailAndPassword,
  signOut,
  onAuthStateChanged,
  User as FirebaseUser,
} from 'firebase/auth';
import { doc, getDoc } from 'firebase/firestore';
import { auth, db } from './firebase';
import { User, LoginCredentials, UserRole } from '@/types/auth';

/**
 * Connexion d'un utilisateur avec email et mot de passe
 */
export const login = async (credentials: LoginCredentials): Promise<User> => {
  try {
    const userCredential = await signInWithEmailAndPassword(
      auth,
      credentials.email,
      credentials.password
    );

    const user = await getUserProfile(userCredential.user.uid);
    return user;
  } catch (error) {
    console.error('Erreur lors de la connexion:', error);
    throw new Error('Email ou mot de passe incorrect');
  }
};

/**
 * Déconnexion de l'utilisateur
 */
export const logout = async (): Promise<void> => {
  try {
    await signOut(auth);
  } catch (error) {
    console.error('Erreur lors de la déconnexion:', error);
    throw new Error('Erreur lors de la déconnexion');
  }
};

/**
 * Récupère le profil utilisateur depuis Firestore
 */
export const getUserProfile = async (uid: string): Promise<User> => {
  // Profil par défaut basé sur Firebase Auth
  const defaultProfile: User = {
    uid,
    email: auth.currentUser?.email || '',
    displayName: auth.currentUser?.displayName || auth.currentUser?.email?.split('@')[0] || 'Admin',
    role: UserRole.ADMIN,
  };

  try {
    // Essayer de récupérer depuis Firestore avec un timeout
    const userDoc = await Promise.race([
      getDoc(doc(db, 'users', uid)),
      new Promise<null>((_, reject) => 
        setTimeout(() => reject(new Error('Timeout')), 2000)
      )
    ]);

    if (!userDoc || !userDoc.exists()) {
      console.warn('⚠️ Document utilisateur non trouvé, utilisation du profil par défaut');
      return defaultProfile;
    }

    const data = userDoc.data();
    return {
      uid,
      email: data.email || defaultProfile.email,
      displayName: data.displayName || defaultProfile.displayName,
      role: data.role || UserRole.ADMIN,
    };
  } catch (error) {
    // Toute erreur (timeout, réseau, etc.) → utiliser le profil par défaut
    console.warn('⚠️ Firestore inaccessible, utilisation du profil par défaut');
    return defaultProfile;
  }
};

/**
 * Observer l'état d'authentification
 */
export const observeAuthState = (
  callback: (user: User | null) => void
): (() => void) => {
  return onAuthStateChanged(auth, async (firebaseUser: FirebaseUser | null) => {
    if (firebaseUser) {
      // Toujours appeler le callback avec un utilisateur valide
      // getUserProfile ne lance plus d'erreur, il retourne toujours un profil
      const user = await getUserProfile(firebaseUser.uid);
      callback(user);
    } else {
      callback(null);
    }
  });
};

/**
 * Récupère l'utilisateur actuellement connecté
 */
export const getCurrentUser = (): FirebaseUser | null => {
  return auth.currentUser;
};

File: mobile-driver/driver_app/lib/services/auth_service.dart
import 'package:firebase_auth/firebase_auth.dart';
import 'package:cloud_firestore/cloud_firestore.dart';
import '../models/driver.dart';

class AuthService {
  final FirebaseAuth _auth = FirebaseAuth.instance;
  final FirebaseFirestore _firestore = FirebaseFirestore.instance;

  // Stream d'authentification
  Stream<User?> get authStateChanges => _auth.authStateChanges();

  // Utilisateur actuel
  User? get currentUser => _auth.currentUser;

  /// Connexion avec email et mot de passe
  Future<Driver?> signInWithEmailAndPassword({
    required String email,
    required String password,
  }) async {
    try {
      final credential = await _auth.signInWithEmailAndPassword(
        email: email,
        password: password,
      );

      if (credential.user == null) {
        throw Exception('Échec de la connexion');
      }

      // Récupérer les données du chauffeur depuis Firestore
      final driverDoc = await _firestore
          .collection('users')
          .doc(credential.user!.uid)
          .get();

      if (!driverDoc.exists) {
        throw Exception('Profil chauffeur non trouvé');
      }

      final driverData = driverDoc.data()!;

      // Vérifier que c'est bien un chauffeur
      if (driverData['role'] != 'driver') {
        await _auth.signOut();
        throw Exception('Ce compte n\'est pas un compte chauffeur');
      }

      return Driver.fromJson({
        'id': credential.user!.uid,
        ...driverData,
      });
    } on FirebaseAuthException catch (e) {
      throw _handleAuthException(e);
    }
  }

  /// Déconnexion
  Future<void> signOut() async {
    await _auth.signOut();
  }

  /// Récupérer le profil chauffeur actuel
  Future<Driver?> getCurrentDriver() async {
    final user = currentUser;
    if (user == null) return null;

    try {
      final driverDoc = await _firestore.collection('users').doc(user.uid).get();

      if (!driverDoc.exists) return null;

      final driverData = driverDoc.data()!;
      if (driverData['role'] != 'driver') return null;

      return Driver.fromJson({
        'id': user.uid,
        ...driverData,
      });
    } catch (e) {
      print('Erreur lors de la récupération du profil: $e');
      return null;
    }
  }

  /// Gérer les exceptions Firebase Auth
  String _handleAuthException(FirebaseAuthException e) {
    switch (e.code) {
      case 'user-not-found':
        return 'Aucun utilisateur trouvé avec cet email';
      case 'wrong-password':
        return 'Mot de passe incorrect';
      case 'invalid-email':
        return 'Format d\'email invalide';
      case 'user-disabled':
        return 'Ce compte a été désactivé';
      case 'too-many-requests':
        return 'Trop de tentatives. Veuillez réessayer plus tard';
      case 'network-request-failed':
        return 'Erreur réseau. Vérifiez votre connexion internet';
      default:
        return 'Erreur d\'authentification: ${e.message}';
    }
  }
}
